cpp/src/arrow/type_test.cc

```diff
@@ -426,6 +423,142 @@ TEST_F(TestSchema, TestRemoveMetadata) {
   ASSERT_TRUE(new_schema->metadata() == nullptr);
 }
 
+class TestUnifySchemas : public TestSchema {
+ protected:
+  void AssertSchemaEqualsUnorderedFields(const Schema& lhs, const Schema& rhs) {
```

Comment:
Since field ordering is now deterministic, I think both this and the similar helper in table_test.cc should be removed (and the expected schemas and tables constructed with the expected field ordering)

Reply:
I'm reluctant to make the contract about fields that do not belong to the first schema (they are "appended" to the unified schema in the order they appear in the current implementation) because this contract is a bit complex, and I don't see an obvious benefit to keep it.

That being said.. If you feel that the coverage in FieldOrderingSameAsTheFirstSchema and IdenticalSchema is not enough, I can add additional check in other tests to make sure that fields that belong to the first schema are ordered correctly.